### ISSUE TITLE
Random walk on seemingly equivalent directed and undirected graphs gives different results

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1103]
The following is not technically a bug, but it bit me hard in the behind and it's worth at least documenting with a big attached warning sign.

Consider an undirected graph,

![image](https://user-images.githubusercontent.com/1212871/40674535-6d59cbd2-6375-11e8-8ff7-4bd0dd7db899.png)

and a directed version, 

![image](https://user-images.githubusercontent.com/1212871/40674538-73d1a264-6375-11e8-8255-13b35fe235ad.png)

If we do an `igraph_random_walk` on both, the result (i.e. the frequency with which each vertex is visited) will not be the same.

Finally it became clear to me that this is because the adjacency list of vertex `1` is considered to be `1, 1, 2` and *not* `1, 2` in the undirected version.  This interpretation is not wrong, but neither is it the only reasonable interpretation. 

Also, the relationship between the random walk and the adjacency list functions is not immediately obvious for one not familiar with the internal implementation.

Thus, I think it is very important to document this behaviour, both for the adjacency list and for the random walk functions as a big "WARNING": "this is the particular interpretation that igraph uses".

I should also note that the adjacency matrix function uses a different interpretation than the adjacency list (and the myriad functions that use the adjacency list under the hood). The matrix for this graph is considered to be

```
1 1 0
1 0 1
0 1 1
```

and not

```
2 1 0
1 0 1
0 1 2
```

Thus the naive way of trying to build the stochastic matrix from the adjacency matrix, or even the direct `stochastic_matrix` function in R (which I suppose is `igraph_get_stochastic`), give a stochastic matrix that does not correspond to the random walk process implemented by `igraph_random_walk`.

----

What is your opinion on the matter?  Is the current one the best interpretation for `random_walk`?  Should it be changed?  Should the stochastic matrix or adjacency matrix functions be changed to match?  Perhaps just carefully document the current behaviour for several of the functions that are affected by `igraph_adjlist_init` ... (another one is the undirected -> directed conversion function)

**Comment by ntamas** [comment]
I have just stumbled upon another issue with `igraph_modularity_matrix` (#1035) where a similar fix had to be made for the diagonal elements of the modularity matrix (it assumed 1 for a loop edge instead of 2, while it should definitely be 2 for this particular purpose). I think we should strive to be consistent in such matters throughout igraph so I'd vote for treating the diagonals in the adjacency matrix to be equal to 2*x if we have x loop edges on the node. However, you wrote that:

> This interpretation is not wrong, but neither is it the only reasonable interpretation.

so I'm open to alternatives if we can come up with a reasonable explanation about why we are treating things differently here.

**Comment by szhorvat** [comment]
I would be very cautious with forcing the same interpretation over _all_ affected functions.

Mathematica, Sage and networkx all use the self-loop count (not twice the self-loop count) on the diagonal of the adjacency matrix.

It is worth sticking with this interpretation, if not for anything else than just to avoid surprising users and avoid breaking older programs relying on the current behaviour.

There are applications where the natural choice is to use 1 instead of 2 on the diagonal. For example, the `k`th power of the adjacency matrix contains the number of walks of length `k` between pairs of vertices.  Here, the natural choice is 1.  (One may argue that a self-loop may be traversed in either direction, but that would not be consistent with the interpretation of a power of the adj. matrix as the adj. matrix of a multigraph.)

----

There are of course disadvantages as some natural relations will not hold anymore. E.g., the sum of rows of the adj. matrix is no longer equal to the degree.

----

What would be reasonable in my opinion is:

 - Functions that implement concepts commonly used in many different application areas, such as the adjacency matrix or degree, should stay as they are (unless they go straight against the convention).  Explicitly document choices and point out caveats (e.g. that row sums are not the degree).

 - Functions used in specific application areas, such as the stochastic matrix, should be consistent with other related functions, such as random walks. Changes may be needed here.

----

We should think about this and consider applications carefully. I am not that familiar with these functions because most are very basic and I did not expose them IGraph/M.

Types of functions that may be affected:

 - adjacency matrix (conversion in both direction, all kinds: sparse/dense)
 - degrees, strengths
 - adjacency list, incidence list (conversion in both directions)
 - random walks
 - stochastic matrix
 - processes on graphs (e.g. SIR)
 - similarity functions (number of common neighbours and other similar things that can be computed as dot products of rows of the adj. mat.)
 - functions based on linear algebra that use the adj. mat. (eigenvector centrality, page rank, `community_leading_eigenvector`, etc.)

**Comment by vtraag** [comment]
I agree with @szhorvat that different applications may have different natural representations of the self-loops (either counting them once or twice). Forcing a single perspective on all applications is probably not the most useful. However, it should be clearly documented in which case which interpretation of self-loops is used.

**Comment by ntamas** [comment]
Okay guys, you convinced me. But, in that case, you are probably in a better position to judge what behaviour makes sense from an application perspective as I am not into active research any more. So, feel free to implement whatever behaviour you think makes more sense, as long as it is clearly documented.

**Comment by vtraag** [comment]
While working on #1035, it would indeed make sense to have an option to count self-loops in a variety of ways. I'll open a new issue for this.

**Comment by vtraag** [comment]
I propose that we postpone this until 0.9.0. We can then have a proper solution, as suggested in #1193, which should then be implemented in all relevant functions, not just this one.

**Comment by ntamas** [comment]
Agreed.

**Comment by iosonofabio** [comment]
a variant of this problem emerges also in #972 within PRPACK.

For undirected graphs, the algorithm duplicates every edge and then iterates over vertices and their adjacent neighbors hoping to get every edge twice. Turns out self-loops might be there only once, as written in a fat `TODO` comment in the code.

Not a solution, but a reminder that we should get this done by 0.9

**Comment by szhorvat** [comment]
There is no actionable item in this issue. Closing in favour of #1193.
